### ISSUE TITLE
Saving has_many through when record is new can result in insert/update combo (was working in 4.1.1)

### ISSUE THREAD

**Issue by lowjoel** [rails/rails#16085]
This is normally harmless (albeit inefficient) but not when the join table has the foreign key columns marked as null: false.

Using the following minimal reproduce case:

```ruby
# schema.rb
ActiveRecord::Schema.define(version: 20140707090026) do

  create_table "calendar_events", force: true do |t|
    t.string "test", limit: 45
  end

  create_table "calendar_events_participants", force: true do |t|
    t.integer "calendar_event_id", null: false
    t.integer "person_id",         null: false
  end

  create_table "people", force: true do |t|
  end

end

# calendar_events_participant.rb
class CalendarEventsParticipant < ActiveRecord::Base
	belongs_to :calendar_event, class_name: CalendarEvent.name
	belongs_to :person, class_name: Person.name
end

# calendar_event.rb
class CalendarEvent < ActiveRecord::Base
	has_many :calendar_events_participants
	has_many :participants, through: :calendar_events_participants, source: :person

	accepts_nested_attributes_for :calendar_events_participants, allow_destroy: true
end

# person.rb
class Person < ActiveRecord::Base

end

# in a Rails console
Person.create
e = CalendarEvent.new
e.participants = [Person.first]
e.save
```

The same code would work in Rails 4.1.1 but fail in 4.1.2 onwards (I tested using 4.1.2 and 4.1.4.)

Output:
```
# Rails 4.1.1
   (0.1ms)  BEGIN
  SQL (0.1ms)  INSERT INTO `calendar_events` VALUES ()
  SQL (0.1ms)  INSERT INTO `calendar_events_participants` (`calendar_event_id`, `person_id`) VALUES (1, 1)
   (2.3ms)  COMMIT

# Rails 4.1.2 onwards
   (0.1ms)  BEGIN
  SQL (0.2ms)  INSERT INTO `calendar_events` VALUES ()
  SQL (0.3ms)  INSERT INTO `calendar_events_participants` (`person_id`) VALUES (1)
Mysql2::Error: Field 'calendar_event_id' doesn't have a default value: INSERT INTO `calendar_events_participants` (`person_id`) VALUES (1)
   (1.6ms)  ROLLBACK
```

I can get around the problem by removing the null: false constraint in the database. Then, we will see that the INSERT is followed up immediately by an UPDATE:

```
  (0.1ms)  BEGIN
  SQL (0.1ms)  INSERT INTO `calendar_events` VALUES ()
  SQL (0.1ms)  INSERT INTO `calendar_events_participants` (`person_id`) VALUES (1)
  SQL (0.2ms)  UPDATE `calendar_events_participants` SET `calendar_event_id` = 1 WHERE `calendar_events_participants`.`id` = 1
   (5.4ms)  COMMIT
```

I've tried the same code using sqlite3.

**Comment by al2o3cr** [comment]
This is the same issue as reported in #17015. It is fixed on master, but the fix (removing a call to `add_autosave_association_callbacks` in the implementation of `accepts_nested_attributes_for`) needs to be backported to 4.2.

**Comment by sgrif** [comment]
Thanks @al2o3cr I'll make sure I do that today.

**Comment by magikid** [comment]
@sgrif did you ever get this fix back ported to 4.2?

**Comment by sgrif** [comment]
This was backported to 4.2. Unfortunately, the fix doesn't backport cleanly to 4.1. Since 4.1 is only moderately supported, I'm going to close this. Anyone experiencing this issue should upgrade to 4.2.

**Comment by lowjoel** [comment]
Thanks!

**Comment by sgrif** [comment]
Looks like the fix does actually work on 4-1-stable, but we have a state leak causing the suite to fail when run as a whole I'll see if I can resolve

**Comment by sgrif** [comment]
This fix (and associated problems the fix caused on master) have been backported to 4-1-stable in https://github.com/rails/rails/commit/c3975349270f6a667836bf4cb85e42ec00b341f3, https://github.com/rails/rails/commit/77f6b109061e91eb87e960996e0cb3eef71e3355, and https://github.com/rails/rails/commit/a078cbd95980bb024db63dd7067647fe92384bd4. I believe that 4.2.3 already includes the fixes. They will be released for the 4.1 line in 4.1.13

**Comment by yivo** [comment]
Still buggy on Rails 4.2.3

```ruby
class Profession < ActiveRecord::Base
  has_many :race_to_professions
  has_many :races, through: :race_to_professions
end

class Race < ActiveRecord::Base
  has_many :race_to_professions
  has_many :professions, through: :race_to_professions
end

class RaceToProfession < ActiveRecord::Base
  belongs_to :profession
  belongs_to :race

  def self.table_name
    :races_professions
  end
end
```

Create new record and assign association:
```ruby
p = Profession.new
p.races = [Race.first]
p.save
````
SQL:
```sql
SELECT  `races`.* FROM `races`  ORDER BY `races`.`id` ASC LIMIT 1
INSERT INTO `professions` (`created_at`, `updated_at`) VALUES ('2015-07-31 15:34:18', '2015-07-31 15:34:18')
INSERT INTO `races_professions` (`race_id`, `profession_id`, `created_at`, `updated_at`) VALUES (1, 1, '2015-07-31 15:34:18', '2015-07-31 15:34:18')
```
This works as expected.

Fetch persisted record and assign association:
```ruby
p = Profession.new
p.save
p.races = [Race.first]
p.save
````
SQL:
```sql
INSERT INTO `professions` (`created_at`, `updated_at`) VALUES ('2015-07-31 15:42:02', '2015-07-31 15:42:02')
SELECT  `races`.* FROM `races`  ORDER BY `races`.`id` ASC LIMIT 1
SELECT `races`.* FROM `races` INNER JOIN `races_professions` ON `races`.`id` = `races_professions`.`race_id` WHERE `races_professions`.`profession_id` = 2
INSERT INTO `races_professions` (`race_id`, `created_at`, `updated_at`) VALUES (1, '2015-07-31 15:42:09', '2015-07-31 15:42:09')
```
This does not work as expected.
Examine the query:
```sql
INSERT INTO `races_professions` (`race_id`, `created_at`, `updated_at`) VALUES (1, '2015-07-31 15:42:09', '2015-07-31 15:42:09')
``` 
The column `profession_id` is not updating...
